server/resp/src/main/java/org/infinispan/server/resp/commands/tx/WATCH.java

```diff
@@ -0,0 +1,97 @@
+package org.infinispan.server.resp.commands.tx;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletionStage;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import org.infinispan.AdvancedCache;
+import org.infinispan.commons.util.concurrent.CompletableFutures;
+import org.infinispan.notifications.Listener;
+import org.infinispan.notifications.cachelistener.annotation.CacheEntryCreated;
+import org.infinispan.notifications.cachelistener.annotation.CacheEntryExpired;
+import org.infinispan.notifications.cachelistener.annotation.CacheEntryModified;
+import org.infinispan.notifications.cachelistener.event.CacheEntryEvent;
+import org.infinispan.server.resp.Consumers;
+import org.infinispan.server.resp.Resp3Handler;
+import org.infinispan.server.resp.RespCommand;
+import org.infinispan.server.resp.RespErrorUtil;
+import org.infinispan.server.resp.RespRequestHandler;
+import org.infinispan.server.resp.commands.Resp3Command;
+import org.infinispan.server.resp.commands.TransactionResp3Command;
+import org.infinispan.server.resp.filter.EventListenerKeysFilter;
+import org.infinispan.server.resp.tx.RespTransactionHandler;
+
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.util.AttributeKey;
+
+/**
+ * `<code>WATCH key [key ...]</code>` command.
+ * <p>
+ * Installs a clustered listener to watch for the given <code>key</code>s. The listener receives events for creation,
+ * updates, and expiration.
+ * <p>
+ * The watch instance is local to a single {@link ChannelHandlerContext}. To remove the listeners, the same context
+ * needs to execute the operation. There is no way to remove a single specific watcher. All listeners deregister during
+ * an {@link EXEC}, {@link UNWATCH}, DISCARD, transaction abort, or closed channel.
+ * <p>
+ * Since a listener is bound to a single connection, this ensures that another client does not affect each other's
+ * transactions safeguards.
+ *
+ * @since 15.0
+ * @see <a href="https://redis.io/commands/watch/>Redis Documentation</a>
+ * @author José Bolina
+ */
+public class WATCH extends RespCommand implements Resp3Command, TransactionResp3Command {
+
+   static final AttributeKey<List<TxKeysListener>> WATCHER_KEY = AttributeKey.newInstance("watchers");
+
+   private static final byte[][] EMPTY = new byte[0][];
+
+   public WATCH() {
+      super(-2, 1, -1, 1);
+   }
+
+   @Override
+   public CompletionStage<RespRequestHandler> perform(Resp3Handler handler, ChannelHandlerContext ctx, List<byte[]> arguments) {
+      AdvancedCache<byte[], byte[]> cache = handler.cache();
+      TxKeysListener listener = new TxKeysListener();
+      byte[][] keys = arguments.toArray(EMPTY);
+
+      CompletionStage<Void> cs = cache.addListenerAsync(listener, new EventListenerKeysFilter(keys, cache.getKeyDataConversion()), null)
```

Comment:
I would suggest using a Converter to reduce the payload to a smaller size, there is no reason to send the entire value payload to a remote client when all it is doing is marking the tx as having a modification.